### DISCUSSION TITLE
Feature request: AreTomo2 integration

### DISCUSSION THREAD

**Discussion by alexjnoble** [nextpyp/.github#15] [Ideas]
Hello,

I think it would be nice for users to have access to multiple tilt-series alignment methods. One that we use and is quite fast (~2 minutes per tilt-series per GPU) is AreTomo2:

https://github.com/czimaginginstitute/AreTomo2

AreTomo2 outputs files in IMOD format, so it should be pretty easy to integrate. Alignment is done using iterative projection matching, similar to Protomo, which we find works in most cases. There is also CTF estimation, but we haven't analyzed that yet.

AreTomo2 is BSD-3, so integration shouldn't be a problem. There is no Singularity build, so that would need to be created if necessary. 

Thank you,
-Alex

**Comment by asarnow** [comment]
AreTomo has very robust (much more so than IMOD) determination of the tilt axis and 2D shifts without fiducials, but doesn't refine tilt angles and seems to have worse Z-resolution. AreTomo can write output for IMOD - it would be very cool if PYP could use the strengths of both.

**Comment by abartesaghi** [comment]
We are yet to see cases where IMOD fiducial-based or patch-tracking fail to align tilt-series, but we see the benefit of having different options for tilt-series alignment so we agree that integrating AreTomo2 into nextPYP would be a good idea. Another consideration is that because AreTomo2 uses GPUs, this may have implications for scalability (since many clusters have comparatively fewer GPUs than CPUs available these days) and compatibility (since it would not be possible to fully containerize the CUDA-dependent code the same way all other software in nextPYP is containerized). Overall, providing more options for tilt-series alignment will probably outweigh these limitations.

**Comment by asarnow** [comment]
I have quite a few examples where IMOD is not doing alignment well (but only absent fiducials), and vice versa with AreTomo which seems sometimes to be disrupted by the presence of fiducials. Perhaps my samples are poor.

Also, the only thing that can't be containerized is the NVIDIA driver itself. The CUDA libraries the program is built against can be included in the container. The additional requirement for users would just be that the relevant jobs could only be run on nodes with a sufficiently recent NVIDIA driver installed (min. version for the CUDA version), which is very reasonable IMO.

**Comment by abartesaghi** [comment]
Support for AreTomo2 is available starting with version [0.6.0](https://nextpyp.app/files/pyp/0.6.0/docs/index.html).
